volatility/framework/plugins/windows/netlist.py

```diff
@@ -0,0 +1,544 @@
+
+                # For UdpA, the state is always blank and the remote end is asterisks
+                for ver, laddr, _ in netw_obj.dual_stack_sockets():
+                    yield (0, (format_hints.Hex(netw_obj.vol.offset), "UDP" + ver, laddr, netw_obj.Port, "*", 0, "",
+                               netw_obj.get_owner_pid() or renderers.UnreadableValue(), netw_obj.get_owner_procname()
+                               or renderers.UnreadableValue(), netw_obj.get_create_time()
+                               or renderers.UnreadableValue()))
+
+            elif isinstance(netw_obj, network._TCP_ENDPOINT):
+                vollog.debug("Found _TCP_ENDPOINT @ 0x{:2x}".format(netw_obj.vol.offset))
+                if netw_obj.get_address_family() == network.AF_INET:
+                    proto = "TCPv4"
+                elif netw_obj.get_address_family() == network.AF_INET6:
+                    proto = "TCPv6"
+                else:
+                    proto = "TCPv?"
+
+                try:
+                    state = netw_obj.State.description
+                except ValueError:
+                    state = renderers.UnreadableValue()
+
+                yield (0, (format_hints.Hex(netw_obj.vol.offset), proto, netw_obj.get_local_address()
+                           or renderers.UnreadableValue(), netw_obj.LocalPort, netw_obj.get_remote_address()
+                           or renderers.UnreadableValue(), netw_obj.RemotePort, state, netw_obj.get_owner_pid()
+                           or renderers.UnreadableValue(), netw_obj.get_owner_procname() or renderers.UnreadableValue(),
+                           netw_obj.get_create_time() or renderers.UnreadableValue()))
+
+            # check for isinstance of tcp listener last, because all other objects are inherited from here
+            elif isinstance(netw_obj, network._TCP_LISTENER):
+                vollog.debug("Found _TCP_LISTENER @ 0x{:2x}".format(netw_obj.vol.offset))
+
+                # For TcpL, the state is always listening and the remote port is zero
+                for ver, laddr, raddr in netw_obj.dual_stack_sockets():
+                    yield (0, (format_hints.Hex(netw_obj.vol.offset), "TCP" + ver, laddr, netw_obj.Port, raddr, 0,
+                               "LISTENING", netw_obj.get_owner_pid() or renderers.UnreadableValue(),
+                               netw_obj.get_owner_procname() or renderers.UnreadableValue(), netw_obj.get_create_time()
+                               or renderers.UnreadableValue()))
+            else:
+                # this should not happen therefore we log it.
+                vollog.debug("Found network object unsure of its type: {} of type {}".format(netw_obj, type(netw_obj)))
+
+    def generate_timeline(self):
+        for row in self._generator():
+            _depth, row_data = row
+            # Skip network connections without creation time
+            if not isinstance(row_data[9], datetime.datetime):
```

Comment:
I've added a dictionary if that helps? Still dependent on the order but it's at least indexed by a title now.

```python
    def generate_timeline(self):
        for row in self._generator():
            _depth, row_data = row
            row_dict = {}
            row_dict["Offset"], row_dict["Proto"], row_dict["LocalAddr"], row_dict["LocalPort"], \
            row_dict["ForeignAddr"], row_dict["ForeignPort"], row_dict["State"], \
            row_dict["PID"], row_dict["Owner"], row_dict["Created"] = row_data

            # Skip network connections without creation time
            if not isinstance(row_dict["Created"], datetime.datetime):
                continue
            row_data = [
                "N/A" if isinstance(i, renderers.UnreadableValue) or isinstance(i, renderers.UnparsableValue) else i
                for i in row_data
            ]
            description = "Network connection: Process {} {} Local Address {}:{} " \
                          "Remote Address {}:{} State {} Protocol {} ".format(row_dict["PID"], row_dict["Owner"],
                                                                              row_dict["LocalAddr"], row_dict["LocalPort"],
                                                                              row_dict["ForeignAddr"], row_dict["ForeignPort"],
                                                                              row_dict["State"], row_dict["Proto"])

            yield (description, timeliner.TimeLinerType.CREATED, row_dict["Created"])
```

Reply:
And another thing @ikelos about timeliner. I think this is great plugin, but when running it with `-vvvv` I see that it scans the layer very often to look for `nt_symbols`:
```
Level 9  volatility.framework.configuration.requirements: Symbol table requirement not yet fulfilled: plugins.Timeliner.NetScan.nt_symbols
Level 9  volatility.framework.configuration.requirements: Symbol table requirement not yet fulfilled: plugins.Timeliner.NetScan.nt_symbols
Level 9  volatility.framework.configuration.requirements: Symbol table requirement not yet fulfilled: plugins.Timeliner.NetScan.nt_symbols
DEBUG    volatility.framework.automagic.pdbscan: Kernel base determination - searching layer module list structure
Level 9  volatility.framework.configuration.requirements: Symbol table requirement not yet fulfilled: plugins.Timeliner.DllList.nt_symbols
Level 9  volatility.framework.configuration.requirements: Symbol table requirement not yet fulfilled: plugins.Timeliner.DllList.nt_symbols
Level 9  volatility.framework.configuration.requirements: Symbol table requirement not yet fulfilled: plugins.Timeliner.DllList.nt_symbols
DEBUG    volatility.framework.automagic.pdbscan: Kernel base determination - searching layer module list structure
```
and so on. I do not completely understand how Timeliner works yet but can this be optimized somehow? Maybe with some kind of sharing of nt_symbols?

Reply:
Yeah, that's much nicer!  5:)

So unfortunately timeliner runs multiple different plugins.  At the moment, we don't know what the automagic plugins might fill in (so far, it's usually just primary and nt_symbols, which are pretty common, but we can't guarantee those requirements will even be called that, so we need to go through the automagic every time unfortunately.  I'm not keen on locking down the requirements that timerliner plugins must have, because that'll essentially force people to work a particular way all for the sake of a single plugin.  We *could* theoretically hand the same context through to each plugin (so with layers already made, and transfer the config across *assuming* that they share names, the automagic would fill in the blanks, so should be quick if there's nothing new needed), but I'd definitely put that under a flag to explain what was going on, and it would technically allow a malicious plugin to muddy the context before the next plugin gets given it...  I dunno, might need some thinking about, but happy to have it added as potential idea to the issue tracker?

Reply:
Yeah not sure, don't know how much effort we'd save by transferring over the context. Might be a thing to keep in mind and test at some point.

Regarding potentially malicious plugins: I think we should consider the threat model here - if a malicious plugin is installed, what use case would it have to manipulate the context? It needs user context to install itself into vol's directory at which point it could manipulate vol's core source code too. Furthermore it would be able to manipulate the context of other plugin's only if called via `Timeliner` and not if the other plugins are called directly (except of course the core code is altered too...). Not so sure about this attack vector, but maybe I'm just not creative enough :D

And maybe the base context or base symbol tables or whatever could be set to read-only once they've been created for the first time. Of course, that read-only lock would only be set via user-run code, so a user-run malicious plugin could circumvent that too probably...

Reply:
Well, ok, malicious is probably the wrong word.  I've seen people write _misguided_ plugins that think the easiest way of fixing things they want to have happen would be to just tinker with the context for their own purposes (I'll just make it *look* like something or other).  So not necessarily badly intentioned, just not a thing we want to make easy for people to code lazily.

I think there'd be a lot of machinery involved in letting people do what they want to the context without messing with it.  What we could probably do is run the automagic over the context and then pass that context on, and re-run the automagic (which would already have a lot of the requirements pre-populated, so shouldn't need to fulfill them again).  I think there's the ability to clone a config, so it would reduce the risk to a malicious (or misguided) automagic (which as you point out, is unlikely), so that might be a nice compromise...  That could save a bunch of time.  I'll look into mocking that up at some point...  5:)